bug: JsonToObjectTransformer throws exception when trying to determine ResolvableType

**Affects Version(s):** 5.2.0

## Bug report
We started to get the following error after updating yo Java 13 and Spring 5.2.0. 

The line the causes the error is https://github.com/spring-projects/spring-integration/blob/master/spring-integration-core/src/main/java/org/springframework/integration/json/JsonToObjectTransformer.java#L149

In this case the header value is String version of the target type and not ResolvableType instance.

This is what we use to send the message:
```java
   @Bean
    public IntegrationFlow fileUploadFlow() {
        return IntegrationFlows.from("fileUploadChannel")
                .enrichHeaders(s -> s.headerExpressions(c ->
                        c.put("file-type", "payload.fileUpload.fileType")))
                .transform(Transformers.toJson())
                .handle(Amqp.outboundAdapter(amqpTemplate)
                        .exchangeName(FILE_UPLOAD_EXCHANGE)
                        .routingKey(FILE_UPLOAD_PROCESS_ROUTING_KEY))
                .get();
    }
``` 

And this is the code used to process the message:

```java
    @Bean
    public IntegrationFlow fileUploadFlow(@Qualifier("fileUploadListenerContainer") final SimpleMessageListenerContainer fileUploadListenerContainer) {
        return IntegrationFlows.from(Amqp.inboundAdapter(fileUploadListenerContainer))
                .log()
                .filter("headers['x-death'] == null || headers['x-death'][0].count < 3", f -> f.discardChannel("nullChannel"))
                .transform(Transformers.fromJson(FileUploadRequest.class))
                .routeToRecipients(r ->
                        r.recipientFlow("headers['file-type'].startsWith('image/')", imageFileUploadFlow())
                                .defaultOutputChannel("nullChannel"))
                .get();
    }
```

```
2019-10-31 17:13:11.714  WARN 9674 --- [enerContainer-2] s.a.r.l.ConditionalRejectingErrorHandler : Execution of Rabbit message listener failed.

org.springframework.amqp.rabbit.support.ListenerExecutionFailedException: Listener threw exception
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.wrapToListenerExecutionFailedExceptionIfNeeded(AbstractMessageListenerContainer.java:1683)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1573)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.actualInvokeListener(AbstractMessageListenerContainer.java:1488)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:1476)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:1467)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.executeListener(AbstractMessageListenerContainer.java:1411)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.doReceiveAndExecute(SimpleMessageListenerContainer.java:958)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.receiveAndExecute(SimpleMessageListenerContainer.java:908)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer.access$1600(SimpleMessageListenerContainer.java:81)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.mainLoop(SimpleMessageListenerContainer.java:1279)
	at org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer$AsyncMessageProcessingConsumer.run(SimpleMessageListenerContainer.java:1185)
	at java.base/java.lang.Thread.run(Thread.java:830)
Caused by: org.springframework.integration.transformer.MessageTransformationException: failed to transform message; nested exception is java.lang.IllegalArgumentException: Incorrect type specified for header 'json_resolvableType'. Expected [class org.springframework.core.ResolvableType] but actual type is [class java.lang.String]
	at org.springframework.integration.transformer.AbstractTransformer.transform(AbstractTransformer.java:44)
	at org.springframework.integration.transformer.MessageTransformingHandler.handleRequestMessage(MessageTransformingHandler.java:109)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:127)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:170)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:115)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:133)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:106)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:73)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:453)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:403)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:187)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutput(AbstractMessageProducingHandler.java:444)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.doProduceOutput(AbstractMessageProducingHandler.java:318)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.produceOutput(AbstractMessageProducingHandler.java:266)
	at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutputs(AbstractMessageProducingHandler.java:229)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:133)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:170)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:115)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:133)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:106)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:73)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:453)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:403)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:187)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:166)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:47)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:109)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:198)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter.access$600(AmqpInboundChannelAdapter.java:61)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter$Listener.createAndSend(AmqpInboundChannelAdapter.java:266)
	at org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter$Listener.onMessage(AmqpInboundChannelAdapter.java:232)
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1569)
	... 10 common frames omitted
Caused by: java.lang.IllegalArgumentException: Incorrect type specified for header 'json_resolvableType'. Expected [class org.springframework.core.ResolvableType] but actual type is [class java.lang.String]
	at org.springframework.messaging.MessageHeaders.get(MessageHeaders.java:215)
	at org.springframework.integration.json.JsonToObjectTransformer.obtainResolvableTypeFromHeadersIfAny(JsonToObjectTransformer.java:149)
	at org.springframework.integration.json.JsonToObjectTransformer.doTransform(JsonToObjectTransformer.java:117)
	at org.springframework.integration.transformer.AbstractTransformer.transform(AbstractTransformer.java:33)
	... 44 common frames omitted```

--- a/spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java
+++ b/spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java
@@ -263,13 +263,15 @@ public class DefaultAmqpHeaderMapper extends AbstractHeaderMapper<MessagePropert
 			Map<String, String> jsonHeaders = new HashMap<>();
 
 			for (String jsonHeader : JsonHeaders.HEADERS) {
-				Object value = getHeaderIfAvailable(headers, jsonHeader, Object.class);
-				if (value != null) {
-					headers.remove(jsonHeader);
-					if (value instanceof Class<?>) {
-						value = ((Class<?>) value).getName();
+				if (!JsonHeaders.RESOLVABLE_TYPE.equals(jsonHeader)) {
+					Object value = getHeaderIfAvailable(headers, jsonHeader, Object.class);
+					if (value != null) {
+						headers.remove(jsonHeader);
+						if (value instanceof Class<?>) {
+							value = ((Class<?>) value).getName();
+						}
+						jsonHeaders.put(jsonHeader.replaceFirst(JsonHeaders.PREFIX, ""), value.toString());
 					}
-					jsonHeaders.put(jsonHeader.replaceFirst(JsonHeaders.PREFIX, ""), value.toString());
 				}
 			}
 			amqpMessageProperties.getHeaders().putAll(jsonHeaders);

--- a/spring-integration-amqp/src/test/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapperTests.java
+++ b/spring-integration-amqp/src/test/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapperTests.java
@@ -20,6 +20,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.fail;
 import static org.mockito.Mockito.mock;
 
+import java.util.Collections;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.Map;
@@ -33,7 +34,9 @@ import org.springframework.amqp.core.MessageProperties;
 import org.springframework.amqp.support.AmqpHeaders;
 import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
 import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.core.ResolvableType;
 import org.springframework.http.MediaType;
+import org.springframework.integration.mapping.support.JsonHeaders;
 import org.springframework.messaging.MessageChannel;
 import org.springframework.messaging.MessageHeaders;
 import org.springframework.messaging.support.GenericMessage;
@@ -308,6 +311,18 @@ public class DefaultAmqpHeaderMapperTests {
 		assertThat(headers.get(AmqpHeaders.DELIVERY_MODE)).isNull();
 		assertThat(headers.get("foo")).isEqualTo("bar");
 		assertThat(headers.get("x-death")).isEqualTo("foo");
+	}
+
+	@Test
+	public void jsonHeadersResolvableTypeSkipped() {
+		DefaultAmqpHeaderMapper headerMapper = DefaultAmqpHeaderMapper.outboundMapper();
+		MessageHeaders integrationHeaders =
+				new MessageHeaders(
+						Collections.singletonMap(JsonHeaders.RESOLVABLE_TYPE, ResolvableType.forClass(String.class)));
+		MessageProperties amqpProperties = new MessageProperties();
+		headerMapper.fromHeadersToReply(integrationHeaders, amqpProperties);
+
+		assertThat(amqpProperties.getHeaders()).doesNotContainKeys(JsonHeaders.RESOLVABLE_TYPE);
 	}
 
 }

--- a/spring-integration-core/src/main/java/org/springframework/integration/json/JsonToObjectTransformer.java
+++ b/spring-integration-core/src/main/java/org/springframework/integration/json/JsonToObjectTransformer.java
@@ -144,11 +144,14 @@ public class JsonToObjectTransformer extends AbstractTransformer implements Bean
 		}
 	}
 
-
+	@Nullable
 	private ResolvableType obtainResolvableTypeFromHeadersIfAny(MessageHeaders headers) {
-		ResolvableType valueType = headers.get(JsonHeaders.RESOLVABLE_TYPE, ResolvableType.class);
+		Object valueType = headers.get(JsonHeaders.RESOLVABLE_TYPE);
+		if (!(valueType instanceof ResolvableType)) {
+			return null;
+		}
 		Object typeIdHeader = headers.get(JsonHeaders.TYPE_ID);
-		if (valueType == null && typeIdHeader != null) {
+		if (typeIdHeader != null) {
 			Class<?> targetClass = getClassForValue(typeIdHeader);
 			Class<?> contentClass = null;
 			Class<?> keyClass = null;
@@ -163,8 +166,7 @@ public class JsonToObjectTransformer extends AbstractTransformer implements Bean
 
 			valueType = JsonObjectMapper.buildResolvableType(targetClass, contentClass, keyClass);
 		}
-
-		return valueType;
+		return (ResolvableType) valueType;
 	}
 
 	private Class<?> getClassForValue(Object classValue) {

--- a/spring-integration-core/src/main/java/org/springframework/integration/json/JsonToObjectTransformer.java
+++ b/spring-integration-core/src/main/java/org/springframework/integration/json/JsonToObjectTransformer.java
@@ -147,9 +147,6 @@ public class JsonToObjectTransformer extends AbstractTransformer implements Bean
 	@Nullable
 	private ResolvableType obtainResolvableTypeFromHeadersIfAny(MessageHeaders headers) {
 		Object valueType = headers.get(JsonHeaders.RESOLVABLE_TYPE);
-		if (!(valueType instanceof ResolvableType)) {
-			return null;
-		}
 		Object typeIdHeader = headers.get(JsonHeaders.TYPE_ID);
 		if (typeIdHeader != null) {
 			Class<?> targetClass = getClassForValue(typeIdHeader);
@@ -166,7 +163,9 @@ public class JsonToObjectTransformer extends AbstractTransformer implements Bean
 
 			valueType = JsonObjectMapper.buildResolvableType(targetClass, contentClass, keyClass);
 		}
-		return (ResolvableType) valueType;
+		return valueType instanceof ResolvableType
+				? (ResolvableType) valueType
+				: null;
 	}
 
 	private Class<?> getClassForValue(Object classValue) {

